### PR TITLE
v8 Filter upgrade + tidy up 

### PR DESCRIPTION
**Pull request by GoodBoyDigital**
- introduced rendererType enum
- removed globalUniforms from filters, added `outputTexture` instead
- updated all filters to work with new filter layout
- added compatibleRenderers option to Filter
- fixed multi-pass filters
- use correct stencil type for webGL2
- added cache to setting a texture id on a shader in GLShaderSystem
- added a listener to when a texture resizes in WebGL
- if a generic object resource is passed in, it is converted to a uniform group
- fixed an offset issue when uploading an array to a uniform buffer
- fixed texturePool `getSameSizeTexture` bug
- fixed getGlobalBounds bug
- fixed mask + blendMode bug
- added Blur Filters

### Summary
🎨🐛🔧

This pull request refactors the filter system to use the `outputTexture` uniform instead of the `backgroundFrame` uniform for vertex and texture calculations, simplifies the uniform groups and bindings, and improves the compatibility and performance of the filters with the webgl2 renderer. It also fixes some bugs with the displacement and mask filters, adds the BlurFilter and BlurFilterPass classes, and updates some constants and imports.

> _Sing, O Muse, of the glorious deeds of the web developers_
> _Who refactored the filter system with skill and wisdom_
> _They changed the `outputTexture` uniform to simplify the shaders_
> _And improved the performance and flexibility of the renderers_

### Walkthrough
*  Refactor the filter system to use the outputTexture uniform instead of the backgroundFrame uniform to calculate the vertex position and texture coordinates for the filters, and to avoid un-premultiplying and re-premultiplying the alpha of the input and output colors for the filters ([link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-5b9422543e89171bde9a0937a5ca58baaffb8faf039b09c81d2f77923d864f6dL12-R13),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-5b9422543e89171bde9a0937a5ca58baaffb8faf039b09c81d2f77923d864f6dL21-R26),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-5b9422543e89171bde9a0937a5ca58baaffb8faf039b09c81d2f77923d864f6dL37-R41),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-5b9422543e89171bde9a0937a5ca58baaffb8faf039b09c81d2f77923d864f6dL70-R78),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-5b9422543e89171bde9a0937a5ca58baaffb8faf039b09c81d2f77923d864f6dL108-R113),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-23df363d21b6df225d0b13abdb73cb151a3bfc8b1eeadd8d4d2249da11fdff0cL4-R18),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-198215ea87b61cf8297db136aa9817ace442eb22128e39f451d87c6456f79626L5-R9),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-198215ea87b61cf8297db136aa9817ace442eb22128e39f451d87c6456f79626L18-R19),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-afe2037cdd58c776f061f947876e751e918e16eb53caa9a1595392a044d9f91aL12-R13),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-afe2037cdd58c776f061f947876e751e918e16eb53caa9a1595392a044d9f91aL24-R45),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-afe2037cdd58c776f061f947876e751e918e16eb53caa9a1595392a044d9f91aL56-L60),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-abbbd0db1cc320efcc4ab6ab94e47afe9eed455fd6447f00aed468023d514aabL12-R9),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-abbbd0db1cc320efcc4ab6ab94e47afe9eed455fd6447f00aed468023d514aabL21-R26),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-abbbd0db1cc320efcc4ab6ab94e47afe9eed455fd6447f00aed468023d514aabL42-R36),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-abbbd0db1cc320efcc4ab6ab94e47afe9eed455fd6447f00aed468023d514aabL50-L54),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-abbbd0db1cc320efcc4ab6ab94e47afe9eed455fd6447f00aed468023d514aabL71-R60),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-abbbd0db1cc320efcc4ab6ab94e47afe9eed455fd6447f00aed468023d514aabL86),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-abbbd0db1cc320efcc4ab6ab94e47afe9eed455fd6447f00aed468023d514aabL94),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-30ec6a0d5027145c04df5d4168dbfe300a33450684cc49e0b066652195eed731L3),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-30ec6a0d5027145c04df5d4168dbfe300a33450684cc49e0b066652195eed731L17-R16),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-96b09cb6be04b4032bcf3ca5cb49e2ed2a5f632e1ce3a42288dfa4e05b7e225bL14-R23),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-96b09cb6be04b4032bcf3ca5cb49e2ed2a5f632e1ce3a42288dfa4e05b7e225bL26-R34),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-07beb33b9099fbd8e20a2fde7ec69aba065044b645b2ad6626fcfdd08fd4fa1fL12-R13),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-07beb33b9099fbd8e20a2fde7ec69aba065044b645b2ad6626fcfdd08fd4fa1fL20-R30),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-07beb33b9099fbd8e20a2fde7ec69aba065044b645b2ad6626fcfdd08fd4fa1fL40-R40),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-07beb33b9099fbd8e20a2fde7ec69aba065044b645b2ad6626fcfdd08fd4fa1fL48-L52),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-07beb33b9099fbd8e20a2fde7ec69aba065044b645b2ad6626fcfdd08fd4fa1fL64-R59),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-07beb33b9099fbd8e20a2fde7ec69aba065044b645b2ad6626fcfdd08fd4fa1fL73-R71),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-cf18ac5e25eb99382cb5914f6c1a2e7392bb8d17ec45616fb4e56af7f6bf6a5dL6-R19),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-629e4b501871e1666fa6c84e4f5ba7adf04e82673224f1efddf7cde31c680635L12-R7),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-629e4b501871e1666fa6c84e4f5ba7adf04e82673224f1efddf7cde31c680635L23-R27),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-629e4b501871e1666fa6c84e4f5ba7adf04e82673224f1efddf7cde31c680635L44-R37),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-629e4b501871e1666fa6c84e4f5ba7adf04e82673224f1efddf7cde31c680635L52-L56),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-629e4b501871e1666fa6c84e4f5ba7adf04e82673224f1efddf7cde31c680635L107-R97),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-383d88a774bfff7f46a4805cdc74a0776e88daf4c21dd3d8959fe6b9615e0dceL96-R93),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-8040a574f56ede509ce32aed02c625d84c94550856acfeff39cc89bd4a80c256L106-R106),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-8040a574f56ede509ce32aed02c625d84c94550856acfeff39cc89bd4a80c256L295-R293),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-8040a574f56ede509ce32aed02c625d84c94550856acfeff39cc89bd4a80c256L324-L343),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-8040a574f56ede509ce32aed02c625d84c94550856acfeff39cc89bd4a80c256L366-L381),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-8040a574f56ede509ce32aed02c625d84c94550856acfeff39cc89bd4a80c256L396-R343),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-8040a574f56ede509ce32aed02c625d84c94550856acfeff39cc89bd4a80c256L465-R359),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-8040a574f56ede509ce32aed02c625d84c94550856acfeff39cc89bd4a80c256L491-R483))
* Simplify the uniform groups and bindings for the filters by removing the globalUniforms group and resource, and using the filterUniforms resource as a generic resource for any filter-specific uniforms ([link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-5b9422543e89171bde9a0937a5ca58baaffb8faf039b09c81d2f77923d864f6dL21-R26),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-afe2037cdd58c776f061f947876e751e918e16eb53caa9a1595392a044d9f91aL24-R45),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-abbbd0db1cc320efcc4ab6ab94e47afe9eed455fd6447f00aed468023d514aabL21-R26),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-07beb33b9099fbd8e20a2fde7ec69aba065044b645b2ad6626fcfdd08fd4fa1fL20-R30),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-629e4b501871e1666fa6c84e4f5ba7adf04e82673224f1efddf7cde31c680635L23-R27),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-383d88a774bfff7f46a4805cdc74a0776e88daf4c21dd3d8959fe6b9615e0dceL96-R93),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-8040a574f56ede509ce32aed02c625d84c94550856acfeff39cc89bd4a80c256L366-L381),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-8040a574f56ede509ce32aed02c625d84c94550856acfeff39cc89bd4a80c256L396-R343),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-8040a574f56ede509ce32aed02c625d84c94550856acfeff39cc89bd4a80c256L491-R483))
* Check the compatibility of the filters with the renderer type before enabling them, and use the name and type fields of the systemConfig object to identify the renderer type ([link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-8040a574f56ede509ce32aed02c625d84c94550856acfeff39cc89bd4a80c256L187-R188),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-d0e6fdc2c2945605a9a22b8c066bb07ce0a57db04dcff5789c12ff913d05adceR2),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-d0e6fdc2c2945605a9a22b8c066bb07ce0a57db04dcff5789c12ff913d05adceL75-R77))
* Add the backTexture field to the FilterData interface to store the backTexture that is used by the filters in the filter data object, and use the previousBounds argument to avoid recalculating the previous bounds from the filter stack ([link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-8040a574f56ede509ce32aed02c625d84c94550856acfeff39cc89bd4a80c256R80),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-8040a574f56ede509ce32aed02c625d84c94550856acfeff39cc89bd4a80c256L271-R280),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-8040a574f56ede509ce32aed02c625d84c94550856acfeff39cc89bd4a80c256L465-R359))
* Use the GlProgram.from and GpuProgram.from static methods instead of the new operator to create the glProgram and gpuProgram instances, and use the cache mechanism of the GlProgram and GpuProgram classes to avoid creating duplicate instances of the same programs ([link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-ceaaa1b42510874331f446bcb905d170838ecb9abbfb28874ecc89b9080f5ea7L49-R55))
* Use the nullish coalescing operator (??) instead of the logical or operator (||) to assign the default value of the scale option in the constructor of the DisplacementFilter class, and use the mapSampler field instead of the mapStyle field to set the sampler resource for the mapTexture ([link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-ceaaa1b42510874331f446bcb905d170838ecb9abbfb28874ecc89b9080f5ea7L35-R35),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-ceaaa1b42510874331f446bcb905d170838ecb9abbfb28874ecc89b9080f5ea7L73-R74))
* Use the super method with the options object instead of destructuring the options object and passing the individual fields to the super method in the constructor of the Filter class, and use the blendMode render pipe to set the blend mode of the mask to 'none' in the pushMask and popMask methods of the StencilMaskPipe class ([link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-383d88a774bfff7f46a4805cdc74a0776e88daf4c21dd3d8959fe6b9615e0dceL75-R75),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-64fa83836f89d14f27a7e04699adc6f3f2dd7ae28843385470001dfc52d6d7e2R52),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-64fa83836f89d14f27a7e04699adc6f3f2dd7ae28843385470001dfc52d6d7e2R115))
* Use the gl.DEPTH24_STENCIL8 constant instead of the gl.DEPTH_STENCIL constant to specify the internal format of the renderbuffer in the createDepthStencilBuffer method of the GlRenderTargetSystem class, and check if the uniform value is different from the defaultSyncData.textureCount before setting the uniform value in the syncUniform method of the GlShaderSystem class ([link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-44089baf2a412137289b97057757debedd8e1db280102cc76cfb5a69b712cf7fL509-R509),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-12ce009d2a0d935bfeb03075623943c691443059175bdc053d4c0bcc8aaf6a77L108-R113))
* Add some imports of types and classes related to the BlurFilter and BlurFilterPass classes that are defined in the `src/filters/blur/BlurFilter.ts` file, and define the BlurFilterPass class, which is a subclass of Filter that applies a horizontal or vertical Gaussian blur to an object ([link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-c57012d40a27839a926680f137aec6a4773459aa43f5474947f401ce29455ab6R1-R199),[link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-335818a046886c55537d367ed937cdecb957812c900beddc8a7666fce8f9b619R1-R146))
* Define the GAUSSIAN_VALUES constant, which is an object that maps different kernel sizes to the corresponding Gaussian weights for the blur filter, in the `src/filters/blur/const.ts` file ([link](https://github.com/pixijs/pixijs/pull/9573/files?diff=unified&w=0#diff-2523d0ac63510377ae84768f0bcab77ab34faa015c6189a7c9121bcdeaedaf08R1-R13))

### CODE REVIEW

File: src/filters/defaultFilter.vert [hunk 1]
```diff
@@ -1,19 +1,21 @@
...
-  mat3 projectionMatrix;
-  mat3 worldTransformMatrix;
-  float worldAlpha;
-};
-
 uniform vec4 inputSize;
 uniform vec4 outputFrame;
+uniform vec4 outputTexture;
+// uniform vec4 globalFrame;
+// uniform float flipped;
```
Comment (Zyie) [inline src/filters/defaultFilter.vert:8]: Can be removed?

File: src/rendering/renderers/shared/shader/utils/createUniformBufferSync.ts [hunk 2]
```diff
@@ -165,6 +167,9 @@ export function generateUniformBufferSync(
 
     const fragmentSrc = funcFragments.join('\n');
 
+    // console.log('--------------------------');
+    // console.log(fragmentSrc);
```
Comment (Zyie) [inline src/rendering/renderers/shared/shader/utils/createUniformBufferSync.ts:171]: Can be removed?

File: src/rendering/scene/utils/buildInstructions.ts [hunk 3]
```diff
@@ -19,35 +19,14 @@ export function buildInstructions(layerGroup: LayerGroup, renderPipes: RenderPip
...
-
-    const length = children.length;
-
-    for (let i = 0; i < length; i++)
-    {
-        collectAllRenderables(children[i], instructionSet, renderPipes);
-    }
+    collectAllRenderablesAdvanced(root, instructionSet, renderPipes, true);
 
+    // instructionSet.log();
```
Comment (Zyie) [inline src/rendering/scene/utils/buildInstructions.ts:29]: Can be removed?

File: src/types.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+/* eslint-disable @typescript-eslint/indent */
+/**
```
Comment (Zyie) [inline src/types.ts:2]: Move to src/rendering

**Review by Zyie — APPROVED** [review]

File: src/types.ts (hunk 4 above)
Comment (GoodBoyDigital) [reply src/types.ts:2]: cool, will be done in final PR 👍

**Pull request merged**